### ISSUE TITLE
[Question] Is there a built-in byte[] to hex formatter?

### ISSUE THREAD

**Issue by levicki** [NLog/NLog#5533]
Sorry for posting a question here, I tried searching and I couldn't find anything on SO about this topic.

Assuming code like this:
```
byte[] CmdBytes = new byte[] {
	0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F,
	0x10, 0x11 ,0x12, ...
};
Int32 hr = SendCommand(CmdBytes);
m_Logger.LogInformation("SendCommand(CmdBytes): CmdBytes = {command}, hr = {hr:X8}", CmdBytes, hr);
```
I need the ability to log `byte[]` contents, preferably formatted like this:
```
2024-05-27T14:37:59: SendCommand(CmdBytes): CmdBytes = [
	0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F,
	0x10, 0x11 ,0x12, ...
], hr = 0x8007005
```
I am aware that I could write a helper to preformat the data to string and use that with structured logging but that would be ugly.

Is there perhaps such a formatter available already and if not, what would be the best way to implement custom formatter for this purpose?

Please advise.

**Comment by snakefoot** [comment]
There is no builtin formatter for byte-arrays. I suggest to wrap the byte-array in a struct that implements [IFormatable](https://learn.microsoft.com/en-us/dotnet/api/system.iformattable). Ex:
```c#
struct ByteArrayFormatter :  IFormatable
{
     private readonly byte[] _byteArray;

     string IFormatable.ToString(string format, IFormatProvider formatProvider) => ToString();

     public ByteArrayFormatter(byte[] byteArray)
     {
          _byteArray = byteArray;
     }

     public string ToString()
     {
           // Magic formatting happening here
     }
}
```
And call like this:
```c#
var cmdBytesWrapper = new ByteArrayFormatter(CmdBytes)
m_Logger.LogInformation("SendCommand(CmdBytes): CmdBytes = {command}, hr = {hr:X8}", cmdBytesWrapper, hr);
```

**Comment by levicki** [comment]
How is that better than just calling a method that converts `byte[]` to `string` directly in parameter list?

Isn't there a better way which doesn't require modifying every `byte[]` I want to log everywhere in the code?

What of `IValueFormatter`? Can perhaps that be used?

**Comment by levicki** [comment]
@snakefoot I would like to use something like this:
```
public bool FormatValue(object value, string format, CaptureType captureType, IFormatProvider formatProvider, StringBuilder builder)
{
	if (captureType == CaptureType.Normal) {
		if (value is byte[]) {
			byte[] bytes = (byte[])value;
			builder.Append("[\r\n");
			int Count = bytes.Length / 16;
			int Remainder = bytes.Length % 16;
			for (int i = 0; i < Count; i += 16) {
				builder.Append("\t");
				for (int j = 0; j < 16; j++) {
					builder.Append($"0x{bytes[(i * 16) + j]:X2}");
					if (j != 16 - 1) {
						builder.Append(", ");
					} else {
						builder.Append(",\r\n");
					}
				}
			}
			builder.Append("\t");
			for (int k = 0; k < Remainder; k++) {
				builder.Append($"0x{bytes[(Count * 16) + k]:X2}");
				if (k != Remainder - 1) {
					builder.Append(", ");
				} else {
					builder.Append("\r\n");
				}
			}
			builder.Append("]");

			return true;
		}
	}

	return false;
}
```
Also, I need a `ILoggerFactory` so I can use it with `Microsoft.Extensions.Logging` and custom DI code. I was obtaining a factory by using `new NLogLoggerFactory()`, however I have no idea how to register this formatter to it.

Finally, registering this formatter does not seem to chain it with default formatter so if I don't handle all object types myself instead of just the ones I want to override then structured logging is practically broken for all other object types except this one.

Please advise.

**Comment by snakefoot** [comment]
The [NLog Dependency Injection](https://github.com/NLog/NLog/wiki/Dependency-injection-with-NLog) overrides should happen before having built the `LoggingConfiguration`. Ex:
```c#
            LogManager.Setup().SetupExtensions(ext => {
                var defaultFormatter = ext.LogFactory.ServiceRepository.GetService(typeof(IValueFormatter));
                var newFormatter = new CustomValueFormatter((IValueFormatter)defaultFormatter);
                ext.RegisterSingletonService<IValueFormatter>(newFormatter);
            }).LoadConfiguration(cfg => {
                var newConfig = new LoggingConfiguration(cfg.LogFactory);
                cfg.Configuration = newConfig;
            });
```
See also: https://github.com/NLog/NLog/wiki/Fluent-Configuration-API

Example of custom `IValueFormatter`:
```c#
    public class CustomValueFormatter : IValueFormatter
    {
        private readonly IValueFormatter _defaultValueFormatter;
    
        public CustomValueFormatter(IValueFormatter defaultValueFormatter)
        {
            _defaultValueFormatter = defaultValueFormatter;
        }
    
        public bool FormatValue(object value, string format, CaptureType captureType, IFormatProvider formatProvider,
            StringBuilder builder)
        {
            if (captureType == CaptureType.Normal && value is byte[])
            {
                  builder.Append("magic");
                  return true;
            }
            return _defaultValueFormatter.FormatValue(value, format, captureType, formatProvider, builder);
        }
    }
```

**Comment by levicki** [comment]
@snakefoot Thanks for clarification, I was confused by deprecation notices.

However, one thing still isn't clear to me. Namely, I don't understand how to get from:
```
            LogManager.Setup().SetupExtensions(ext => {
                var defaultFormatter = ext.LogFactory.ServiceRepository.GetService(typeof(IValueFormatter));
                var newFormatter = new CustomValueFormatter((IValueFormatter)defaultFormatter);
                ext.RegisterSingletonService<IValueFormatter>(newFormatter);
            }).LoadConfigurationFromFile();
```
To:
```
ILoggerFactory myFactory;
```
I need to pass an `ILoggerFactory` instance into my library which has a lot of classes which use `ILogger<T>` pattern for internal logging. I am not doing this in a service so typical config steps for .Net Core web services / apps don't really work for me.

TL;DR -- how to get an `ILoggerFactory` from `LogManager` after you've configured custom `IValueFormatter`?

**Comment by snakefoot** [comment]
When writing a library/plugin then it becomes a guest in the application-house. It is not nice when the guest replaces the carpet in the living room.

The application hostbuilder has the task of setting up logging-configuration (Ex. setup NLog and call `AddNLog` or `UseNLog`) and other dependency-injection-logic. The `ILoggerFactory` is the result after logging-setup has completed, and should not be modified afterwards.

**Comment by levicki** [comment]
@snakefoot I am afraid you misunderstood me &mdash; I am not trying to override formatter from library.

Application will still setup the logging configuration including the formatter. I just said I don't know how to obtain `ILoggerFactory` reference to pass from the application to the library based on the setup steps you provided.

**Comment by snakefoot** [comment]
I was thinking you had a place where you setup NLog. That would be the place where you setup everything (before building the `ILoggerFactory`).

See also: https://github.com/NLog/NLog/wiki/Getting-started-with-.NET-Core-2---Console-application#a-minimal-example
See also: https://github.com/NLog/NLog/wiki/Getting-started-with-ASP.NET-Core-6

**Comment by levicki** [comment]
@snakefoot As I said this is .Net Framework, not .Net Core or ASP.NET.

Let me give a simplest possible example:
```
namespace MyApp
{
	using Microsoft.Extensions.Logging;
	using NLog.Extensions.Logging;

	internal class Program
	{
		private static ILoggerFactory	m_LoggerFactory;

		private static ILoggerFactory SetupNLog()
		{
			return new NLogLoggerFactory(); // This is how I was creating an ILoggerFactory
		}

		static void Main(string[] args)
		{
			m_LoggerFactory = SetupNLog();

			MyMainClass mc = new MainClass(m_LoggerFactory);
		}
	}
}

namespace MyLib
{
	using Microsoft.Extensions.Logging;

	public class MyMainClass()
	{
		private readonly ILoggerFactory m_Factory;
		private readonly ILogger<MyMainClass> m_Logger;

		public MyMainClass(ILoggerFactory Factory)
		{
			m_Factory = Factory; // Store so it can be used to create ILogger<T> later for any new class added
			m_Logger = m_Factory.CreateLogger<MyMainClass>();
		}
	}
}
```
The problem is that now when I need to add IValueFormatter I don't know how to get ILoggerFactory from it. I hope it's now clear what I want?

**Comment by snakefoot** [comment]
Seems most people uses Microsoft Dependency Injection to setup the Microsoft LoggingFactory, and then register the wanted logging-providers (like NLog). All examples for ASP.NET Core / .NET Core also works with .NET Framework.

When using `new NLogLoggerFactory` (without any parameters) then you are using `NLog.LogManager.LogFactory`. So you can just do like this:
```c#
namespace MyApp
{
	using Microsoft.Extensions.Logging;
	using NLog.Extensions.Logging;
	using MyLib;

	internal class Program
	{
		private static ILoggerFactory	m_LoggerFactory 

		private static ILoggerFactory SetupNLog()
		{
                        // Setup NLog before creating Logger / ILoggerFactory
			LogManager.Setup().SetupExtensions(ext => {
			    var defaultFormatter = ext.LogFactory.ServiceRepository.GetService(typeof(IValueFormatter));
			    var newFormatter = new CustomValueFormatter((IValueFormatter)defaultFormatter);
			    ext.RegisterSingletonService<IValueFormatter>(newFormatter);
			    });
			return new NLogLoggerFactory();
		}

		static void Main(string[] args)
		{
			m_LoggerFactory = SetupNLog();

			MyMainClass mc = new MyMainClass(m_LoggerFactory);
		}
	}
}
```

**Comment by levicki** [comment]
> When using `new NLogLoggerFactory` (without any parameters) then you are using `NLog.LogManager.LogFactory`. So you can just do like this:

I just tried that and the new formatter isn't being called.

**Comment by snakefoot** [comment]
By default NLog just uses the string-formatting provided by Microsoft Extensions Logging. Because Microsoft Extension Logging logging-interface already perform implicit parsing of the message-template (faster). But you can force NLog to perform its own parsing and string-formatting by using `new NLogProviderOptions() { ParseMessageTemplates = true }`:
```c#
```c#
namespace MyApp
{
	using Microsoft.Extensions.Logging;
        using NLog;
	using NLog.Extensions.Logging;
	using MyLib;

	internal class Program
	{
		private static ILoggerFactory	m_LoggerFactory 

		private static ILoggerFactory SetupNLog()
		{
                        // Setup NLog before creating Logger / ILoggerFactory
			LogManager.Setup().SetupExtensions(ext => {
			    var defaultFormatter = ext.LogFactory.ServiceRepository.GetService(typeof(IValueFormatter));
			    var newFormatter = new CustomValueFormatter((IValueFormatter)defaultFormatter);
			    ext.RegisterSingletonService<IValueFormatter>(newFormatter);
			    });
                        // Force NLog String-Formatting, instead of using Microsoft Extension Logging string-formatting
			return new NLogLoggerFactory(new NLogProviderOptions() { ParseMessageTemplates = true });
		}

		static void Main(string[] args)
		{
			m_LoggerFactory = SetupNLog();

			MyMainClass mc = new MyMainClass(m_LoggerFactory);
		}
	}
}
```
```

**Comment by levicki** [comment]
Also to answer this:

> Seems most people uses Microsoft Dependency Injection

If Microsoft made it easy to use their DI from .Net Framework apps (or C# nuget packages from C++/CLR code, but I digress) and if people didn't have to resort to 3rd party solutions or rolling their own for a lot of stuff, wouldn't that be really cool?

**Comment by levicki** [comment]
> But you can force NLog to perform its own parsing and string-formatting by using...

That works, thanks. But what are performance implications? Is it double parsing then?

**Comment by snakefoot** [comment]
> But what are performance implications? Is it double parsing then?

Yes, the Microsoft Extension Logging ILogger-extension-methods performs the parsing upfront (before reaching the NLog-provider-logic). NLog cannot perform string-formatting without doing its own parsing.

I guess you can change to this logic instead (Notice the `@`-prefix), and then change custom `IValueFormatter` to `if (captureType == CaptureType.Serialize && value is byte[] byteArray)`:
```c#
logger.LogError("{@World}", System.Text.Encoding.UTF8.GetBytes("World"));
```
NLog will automatically perform additonal parsing when meeting properties with `@`-prefix.

**Comment by levicki** [comment]
> Yes, the Microsoft Extension Logging ILogger-extension-methods performs the parsing upfront (before reaching the NLog-provider-logic). NLog cannot perform string-formatting without doing its own parsing.

I guess it would have been too much to ask of Microsoft developers to design logging extensions which everyone will use and have to piggy-back on in such a way that you can register callbacks or override methods or something. I mean, even back in the DOS and assembly days one could install custom handlers for stuff, not seeing what is stopping us today with this much more advanced coding knowledge.

> Could be nice if...

I totally agree.

> ...when not wanting to use...

I am not a fan of more typing (i.e. replacing all `byte[]` with wrappers in code), I prefer "one and done" approach.

In any case, thanks for the help even if it's double parsing in the end at least it does what I want.

Perhaps you decide to include the formatter in the NLog at some point and I won't need it anymore. If you decide to do so, based on the code I posted make a note that size check should be added and limit to dumping first `n` bytes.
